### PR TITLE
Use index files in 2L nav

### PR DESCRIPTION
**Pull request by piebrain**
Automatically generates .index file in folder if it is missing. It will then use this index to find filenames from an ID.

There's room for improvements, but it works decently. It's the hope that is can be used in the other navigation schemes as well, but it will probably need a certain level of abstraction to do that.

Comments are very welcome.

### CODE REVIEW

File: lib/Compass/Navigation2L.cpp [hunk 1]
```diff
@@ -46,7 +47,8 @@ uint8_t Navigation2L::up()
     folderIndex--;
     if (folderIndex < 0)
         folderIndex = folderCount-1;  //loop back to the end of the list
-    listFiles();
+    player->stop();
```
Comment (sastel) [inline lib/Compass/Navigation2L.cpp:50]: Why did you have to add a call to player->stop() here and in down()?

File: lib/Compass/Navigation2L.cpp [hunk 2]
```diff
@@ -128,26 +130,12 @@ void Navigation2L::listFolders()
...
-void Navigation2L::listFiles()
+void Navigation2L::indexAllFolders()
 {
-    Dir root = SDFS.openDir(folders[folderIndex]);
-
-    int i = 0;
-    while(root.next())
+    player->stop();
+    listFolders();
+    for (String folder : folders)
```
Comment (sastel) [inline lib/Compass/Navigation2L.cpp:137]: Very elegant! I didn't know this was a feature in C++.

File: lib/Compass/nav/FileIndex.cpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    int len = indexText.length();
+    for (i = 0; i * 5 < len; i++) {
+        index[i] = indexText.substring(i * 5, (i * 5) + 5).toInt();
+    }
+    indexTracks = i;
+
+    return indexTracks;
+}
+
+void FileIndex::createIndexFile(String folder)
```
Comment (sastel) [inline lib/Compass/nav/FileIndex.cpp:29]: It would improve readability if you can abstract some of the logic blocks found in this method into private methods.

File: lib/Compass/Navigation2L.cpp (hunk 2 above)
Comment (piebrain) [reply lib/Compass/Navigation2L.cpp:137]: Since C++11

File: lib/Compass/Navigation2L.cpp (hunk 1 above)
Comment (piebrain) [reply lib/Compass/Navigation2L.cpp:50]: I know it's a bit ugly, but it's to make sure there is enough free memory for creating index files and since I'm creating them on the fly for now if they are missing.

File: lib/Compass/nav/FileIndex.cpp [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+{
+    DEBUG_COMPASS_LIB("Creating Index File for %s\n", folder.c_str());
+    String tracks[INDEXER_MAX_TRACKS];
```
Comment (sastel) [inline lib/Compass/nav/FileIndex.cpp:32]: Do you think it would be better to use a linked list? That way we won't allocate any more space than what we need to store all of the track names that exist at a particular level. It is also true though that use of a linked list will use slightly more memory per element. That can accumulate quickly and may be a problem if there is a very high number of tracks in a folder.

File: lib/Compass/nav/FileIndex.cpp [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    String indexFilename = folder + F(".index");
+    if(!SDFS.exists(indexFilename)) {
+        createIndexFile(folder);
+    }
+
+    indexFile = SDFS.open(indexFilename, "r");
+    uint8_t version = indexFile.readStringUntil('\n').toInt();
+    String indexText = indexFile.readStringUntil('\n');
```
Comment (sastel) [inline lib/Compass/nav/FileIndex.cpp:17]: Depending on the number of tracks, this could be a very large string. Instead of storing the position as a series of characters with 5 chars per position, can we store raw binary? That would be a more efficient use of space.

**Review by sastel — APPROVED** [review]
I left some comments for your consideration.

File: lib/Compass/Navigation2L.cpp (hunk 1 above)
Comment (sastel) [reply lib/Compass/Navigation2L.cpp:50]: That makes sense now that I see how much memory is being used in FileIndex::open()! :)

File: lib/Compass/nav/FileIndex.cpp (hunk 4 above)
Comment (piebrain) [reply lib/Compass/nav/FileIndex.cpp:32]: I'm currently rewriting this to account for indexing tracks, folders and VPs. I'll try my best to use less memory by not having a fixed sized array. A vector would probably be the better choice if you want to go that route.

File: lib/Compass/nav/FileIndex.cpp (hunk 5 above)
Comment (piebrain) [reply lib/Compass/nav/FileIndex.cpp:17]: Very true! I just lack the skills and time to work with binary...

**Comment by piebrain** [comment]
It now indexes all relevant files: Tracks, Folders, VPCs and VP[LSB].

For VP[LSB]s it is assumed there is only 1 of them in a folder and that one can be reached with the getFolderVP() method.

If you are loading a track index and there are no files in the folder, then count will be set to -1 and getFilename() will return an empty string.

A problem I have noticed is that sometimes it is not loading from the index file correctly, which leads to missing the end of a filename. If you go to the next folder and then back everything seems to be OK again.

Memory wise there should be no problems setting INDEXER_MAX_TRACKS to 200 or 250, but when I do it crashes the player while creating the index with some kind of yield error. I have no idea why!

File: lib/Compass/Navigation2L.cpp [hunk 6]
```diff
@@ -106,6 +106,15 @@ void Navigation2L::setTrackData(String data)
...
 
+void Navigation2L::openFolder()
+{
+    if(player != nullptr) {
+        player->stop(); // to ensure memory for indexing
+    }
+    currentFolder = "/" + folders->getFilename(folderIndex) + "/";
+    tracks->open(currentFolder);
+}
+#if 0
```
Comment (sastel) [inline lib/Compass/Navigation2L.cpp:117]: Can this section be removed?

File: lib/Compass/Navigation2L.cpp [hunk 7]
```diff
@@ -116,38 +125,46 @@ void Navigation2L::listFolders()
...
-void Navigation2L::listFiles()
+#endif
+void Navigation2L::indexAllFolders()
 {
-    Dir root = SDFS.openDir(folders[folderIndex]);
-
-    int i = 0;
-    while(root.next())
+    player->stop();
+    /*
```
Comment (sastel) [inline lib/Compass/Navigation2L.cpp:143]: Why is this section commented out?

**Comment by sastel** [comment]
> It now indexes all relevant files: Tracks, Folders, VPCs and VP[LSB].
> 
> For VP[LSB]s it is assumed there is only 1 of them in a folder and that one can be reached with the getFolderVP() method.
> 
> If you are loading a track index and there are no files in the folder, then count will be set to -1 and getFilename() will return an empty string.
> 
> A problem I have noticed is that sometimes it is not loading from the index file correctly, which leads to missing the end of a filename. If you go to the next folder and then back everything seems to be OK again.
> 
> Memory wise there should be no problems setting INDEXER_MAX_TRACKS to 200 or 250, but when I do it crashes the player while creating the index with some kind of yield error. I have no idea why!

Can you share the error text below?
I managed to crash the player with a watchdog timeout when I was developing the Factory Built In Test. I had an infinite loop with no call to yield() or delay(). Could your crash be happening because the indexing routine takes so long to run that the watchdog never gets kicked? Maybe placing a call to yield in there will fix it.

File: lib/Compass/nav/FileIndex.h [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+#define FILEINDEX_VERSION 1
+
+// 150 seems to be golden. Changing it to 200 makes the device panic
+// If the index is build on the computer and loaded to
+// the device, then we can probably load many more
+// Disclaimer: Not tested with 64 char filenames
+#define INDEXER_MAX_TRACKS  150 
+
+#define INDEX_TYPE_TRACK 0
```
Comment (sastel) [inline lib/Compass/nav/FileIndex.h:17]: An enum makes more sense here.

File: lib/Compass/nav/FileIndex.cpp [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+    tmp.close();
+}
+
+void FileIndex::createIndexFile(String folder)
+{
+    if(!folder.endsWith("/")) {
+        folder += "/";
+    }
+    DEBUG_COMPASS_LIB("Creating Index File for %s - Type: %i\n", folder.c_str(), type);
+    String files[INDEXER_MAX_TRACKS];
```
Comment (sastel) [inline lib/Compass/nav/FileIndex.cpp:154]: This will take awhile to allocate (and use a tremendous amount of stack space).

Comment (piebrain) [reply lib/Compass/nav/FileIndex.cpp:154]: Is there a way to put it in heap?

Comment (sastel) [reply lib/Compass/nav/FileIndex.cpp:154]: String* files = new String[INDEXER_MAX_TRACKS] I think. But then you will have to deallocate it before this function returns or you will lose the pointer to the array (memory leak).

Comment (sastel) [reply lib/Compass/nav/FileIndex.cpp:154]: createIndexFile only gets called once, and the player isn't running, so it may be fine to put it on the stack.
I'm wondering if this is what causes the yield related crash when INDEXER_MAX_TRACKS is 250. Do you know?

File: lib/Compass/Navigation2L.cpp (hunk 7 above)
Comment (piebrain) [reply lib/Compass/Navigation2L.cpp:143]: Because listFolders() doesn't work anymore... maybe I need to fix that

File: lib/Compass/nav/FileIndex.cpp (hunk 9 above)
Comment (sastel) [reply lib/Compass/nav/FileIndex.cpp:154]: > This will take awhile to allocate (and use a tremendous amount of stack space).

Let me rephrase this - allocating the memory shouldn't take a lot of time, because it is just moving the stack pointer. However, **if** (and I'm certain) the constructor has to be called for each one of those String objects when it is allocated, then it might take a long time. I'm not knowledgeable enough about C++ to be certain.

File: lib/Compass/nav/FileIndex.h (hunk 8 above)
Comment (piebrain) [reply lib/Compass/nav/FileIndex.h:17]: Will do

File: lib/Compass/Navigation2L.cpp (hunk 6 above)
Comment (piebrain) [reply lib/Compass/Navigation2L.cpp:117]: Will do

**Comment by piebrain** [comment]
> > It now indexes all relevant files: Tracks, Folders, VPCs and VP[LSB].
> > For VP[LSB]s it is assumed there is only 1 of them in a folder and that one can be reached with the getFolderVP() method.
> > If you are loading a track index and there are no files in the folder, then count will be set to -1 and getFilename() will return an empty string.
> > A problem I have noticed is that sometimes it is not loading from the index file correctly, which leads to missing the end of a filename. If you go to the next folder and then back everything seems to be OK again.
> > Memory wise there should be no problems setting INDEXER_MAX_TRACKS to 200 or 250, but when I do it crashes the player while creating the index with some kind of yield error. I have no idea why!
> 
> Can you share the error text below?
> I managed to crash the player with a watchdog timeout when I was developing the Factory Built In Test. I had an infinite loop with no call to yield() or delay(). Could your crash be happening because allocating the space for it takes so long to run that the watchdog never gets kicked?

Here's the output with 250:

Creating Index File for /12 - Colossians/ - Type: 0
Done indexing
Loaded 3 entries
   4.    CHARGING got ENTRY_EVT 

User exception (panic/abort/assert)
--------------- CUT HERE FOR EXCEPTION DECODER ---------------

Panic core_esp8266_main.cpp:133 __yield

>>>stack>>>

ctx: cont
sp: 3ffffe00 end: 40100e37 offset: 0000
3ffffe00:  00000001 3ffe8520 4010049e 9ba5e353  
3ffffe10:  000000fe 00000000 00000000 00000000  
3ffffe20:  00000000 00000000 00000000 3ffeef30  
3ffffe30:  00000000 00000003 4bc6a7f0 00000001  
3ffffe40:  3fff0298 00000000 3fff0294 40215d0e  
3ffffe50:  00000000 00000002 80003333 40215d6b  
3ffffe60:  3fff0298 00000f9d 3fff0294 40215855  
3ffffe70:  3fff0298 00000000 3fff0294 4020fd8a  
3ffffe80:  4020ae54 00000000 3ffffec0 4020b4d6  
3ffffe90:  4020ae54 00000000 3ffffec0 4020140c  
3ffffea0:  20474e49 00000a20 3fff1364 4020ae7e  
3ffffeb0:  03530365 0353038c 038e035f 4020ae66  
3ffffec0:  fffffffd 3ffffed0 00000008 00000000  
3ffffed0:  3fff102c 3ffeef50 00000002 4020aef4  
3ffffee0:  3fff1bac 3ffe8974 473aa200 3fff0c24  
3ffffef0:  3ffe8520 fffffffe 3fff1bac 3fff1314  
3fffff00:  00000000 3fff0ecc 3fff1bac 40201784  
3fffff10:  ffffffff 3ffeec44 3fffff50 4020b230  
3fffff20:  40100fe8 3fff1ba4 0000002c 40201526  
3fffff30:  3ffe8520 fffffffe 3fff1bac 3fff0c24  
3fffff40:  3ffe8520 fffffffe 3fff1bac 402017bf  
3fffff50:  ffffffff 4027679c 3fff1bac 4020ae7e  
3fffff60:  00000011 4027679c 3fff1bac 4020180a  
3fffff70:  fffffffd 3ffe8520 0000000a 3fff0c24  
3fffff80:  3ffe8520 3ffeed14 3ffeee7c 4020c3dc  
3fffff90:  feefeffe feefeffe feefeffe 3ffeeef4  
3fffffa0:  3fffdad0 00000000 3ffeeeb4 40215920  
3fffffb0:  feefeffe feefeffe 3ffe85b4 401012ad  
3fffffc0:  feefeffe 3fffefa0 964b46b3 a753d8ea  
3fffffd0:  b70aa171 bbc58459 05d3af6f 401000c0  
3fffffe0:  4010f000 00000004 40001878 4010f19a  
3ffffff0:  4000044c 00000000 00000000 9fb61929  
40000000:  00000000 00000000 00000000 00000000  
40000010:  46007200 0000fffe 00000000 00000000  
40000020:  00003310 00000000 00000000 00000000  
40000030:  46004100 0000fffe 00000000 00000000  
40000040:  00000000 00000000 00000000 00000000  
40000050:  29ffd112 31613951 e820f695 a0323003  
40000060:  71490338 000003a0 00000000 00000000  
40000070:  46004140 0000fffe 00000000 00000000  
40000080:  00000806 00000000 40000000 2222211f  
40000090:  e0000000 400000f3 4000e328 00000000  
400000a0:  00000000 e400000c 13050013 21002010  
400000b0:  eb30fff5 74303003 2356428c 00620200  
400000c0:  20fff221 612013e7 fff02100 61fff151  
400000d0:  030cfff1 6650027d 00058610 00000000  
400000e0:  00506340 f03d0020 3350f03d 14b3b6c0  
400000f0:  70417470 13673440 506340e6 f6c03350  
40000100:  2000edb3 ffe25100 7220030c 34407020  
40000110:  5050e340 7470c033 f0b3f641 21002030  
40000120:  02acffde 12480238 c2222258 0fb3470c  
40000130:  554b0568 334b0369 46f43347 0000fff9  
40000140:  56fe0356 2000fdd5 002e0500 00000000  
40000150:  00000000 00000000 00000000 00000000  
40000160:  00000000 00000000 00000000 00000000  
40000170:  00000000 00000000 00000000 00000000  
40000180:  00000000 00000000 00000000 00000000  
40000190:  00000000 00000000 00000000 00000000  
400001a0:  00000000 00000000 00000000 00000000  
400001b0:  00000000 00000000 00000000 00000000  
400001c0:  00000000 00000000 00000000 00000000  
400001d0:  00000000 00000000 00000000 00000000  
400001e0:  00000000 00000000 00000000 00000000  
400001f0:  00000000 00000000 00000000 00000000  
40000200:  00000000 00000000 00000000 00000000  
40000210:  00000000 00000000 00000000 00000000  
40000220:  00000000 00000000 00000000 00000000  
40000230:  00000000 00000000 00000000 00000000  
40000240:  00000000 00000000 00000000 00000000  
40000250:  00000000 00000000 00000000 00000000  
40000260:  00000000 00000000 00000000 00000000  
40000270:  00000000 00000000 00000000 00000000  
40000280:  00000000 00000000 00000000 00000000  
40000290:  00000000 00000000 00000000 00000000  
400002a0:  00000000 00000000 00000000 00000000  
400002b0:  00000000 00000000 00000000 00000000  
400002c0:  00000000 00000000 00000000 00000000  
400002d0:  00000000 00000000 00000000 00000000  
400002e0:  00000000 00000000 00000000 00000000  
400002f0:  00000000 00000000 00000000 00000000  
40000300:  00000000 00000000 00000000 00000000  
40000310:  00000000 00000000 00000000 00000000  
40000320:  00000000 00000000 00000000 00000000  
40000330:  00000000 00000000 00000000 00000000  
40000340:  00000000 00000000 00000000 00000000  
40000350:  00000000 00000000 00000000 00000000  
40000360:  00000000 00000000 00000000 00000000  
40000370:  00000000 00000000 00000000 00000000  
40000380:  40000000 3fffdab0 3fffeb30 4000dbf8  
40000390:  3fffc100 40000598 3fffc000 4000dc44  
400003a0:  3fffc200 80000000 3fffc180 4000dc54  
400003b0:  3fffc6f8 5851f42d 4c957f2d 7fffffff  
400003c0:  00000060 00000064 00000068 0000006c  
400003d0:  00000010 00000014 00000018 00000020  
400003e0:  00000024 3fffd650 3fffc208 4000dc3c  
400003f0:  000000ff 0000ff00 00ff0000 ff000000  
40000400:  40404040 40404040 80808080 fefefeff  
40000410:  7f800000 00400000 00ffffff 7ff00000  
40000420:  00080000 001fffff 38000000 d411000c  
40000430:  30032cff 201013e6 ffd26100 77ffd371  
40000440:  060906b6 3677664b 00ba05f8 460041f0  
40000450:  0000fffe 0c2ed2f6 ffce6109 81ffd041  
40000460:  caa1ffce ffcab1ff ba1152e0 833a30b5  
40000470:  0b28558a 70c073a0 0b399346 a2a00549  
40000480:  8329a0c0 020cf00d 0000f00d 122c8159  
40000490:  2003b130 013961e6 41091129 a1799169  
400004a0:  c199b189 e1b9d1a9 61d2f1c9 1161e210  
400004b0:  101261f2 bae10020 03e2f0ff d803e4c0  
400004c0:  10ffc01e d90303c0 063f1631 612221c9  
400004d0:  ffaf2213 f140ffc0 0c00ffb3 91f0f040  
400004e0:  d13022f0 e3f0ffb1 b0ccd013 f01e2cd2  
400004f0:  0ef80062 ffd01ed9 13e4f010 f0002010  
40000500:  0cd80060 31101c28 000dc020 f0ffa5e1  
40000510:  62d003e2 c80ed800 10ddc031 5610ffd0  
40000520:  2122016f c21ec913 e4d00221 0061d013  
40000530:  c0ff9f01 00a01303 40ffd000 c0ff9bc1  
40000540:  2ce2b0cd 10ff201f e110ffe0 2f56ff95  
40000550:  ff94f1f8 f0400d00 2cf291d0 13e3d01f  
40000560:  d02022f0 e0863022 000000ff 1013e620  
40000570:  f00d0020 8a41023d 006270ff 04581428  
40000580:  55301439 13e45010 1013e670 f00d0020  
40000590:  0086320c 00000000 61228159 30122c13  
400005a0:  e62003b1 29013961 69410911 89a17991  
400005b0:  a9c199b1 c9e1b9d1 1061d2f1 f21161e2  
400005c0:  20101261 0303c000 d10060d0 21f2ff71  
400005d0:  d021c913 0cc8a0cf 0f3d012d 0cc01c8c  
400005e0:  c021e800 03e00061 ff700113 000000a0  
400005f0:  48ff7041 22030c04 64322a64 00f00d2b  
40000600:  41ff6d51 6a21ff6d f0c112ff 22006102  
40000610:  61220022 2b
--------------- CUT HERE FOR EXCEPTION DECODER ---------------

Panic core_esp8266_main.cpp:133 __yield
